In [1]:
# importing requirements
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

In [2]:
# we need to read the data as our first step after importing libraries
df = pd.read_csv('heart.csv')

In [3]:
# a sample of our data
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Dataset rows details copied from : https://archive.ics.uci.edu/ml/datasets/Heart+Disease
- age : patient's age in years
- sex : defines gender (1 = male; 0 = female)
- cp : chest pain type
- trestbps : resting blood pressure (in mm Hg on admission to the hospital)
- chol : serum cholestoral in mg/dl
- fbs : (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
- restecg : resting electrocardiographic results
- thalach : maximum heart rate achieved
- exang : exercise induced angina (1 = yes; 0 = no)
- oldpeak : ST depression induced by exercise relative to rest
- slope : the slope of the peak exercise ST segment
- ca : number of major vessels (0-3) colored by flourosopy
- thal : 3 = normal; 6 = fixed defect; 7 = reversable defect
- target : have disease or not (1=yes, 0=no

ca, thal and slope rows are considered as categorical variables because they accept a fixed and limited number of possible values. As a result we need to convert them into dummy variables.

In [4]:
# Creation of dummy variables
cp_dummy = pd.get_dummies(df['cp'], prefix = "cp")
thal_dummy = pd.get_dummies(df['thal'], prefix = "thal")
slope_dummy = pd.get_dummies(df['slope'], prefix = "slope")

In [5]:
# now  ca, thal and slope rows will be dropped and dummy variables replace and repressent them.
df = pd.concat([df, cp_dummy, thal_dummy, slope_dummy], axis = 1)
df = df.drop(columns = ['cp', 'thal', 'slope'])
df.head()

age  sex  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  ca  ...  \
0   63    1       145   233    1        0      150      0      2.3   0  ...   
1   37    1       130   250    0        1      187      0      3.5   0  ...   
2   41    0       130   204    0        0      172      0      1.4   0  ...   
3   56    1       120   236    0        1      178      0      0.8   0  ...   
4   57    0       120   354    0        1      163      1      0.6   0  ...   

   cp_1  cp_2  cp_3  thal_0  thal_1  thal_2  thal_3  slope_0  slope_1  slope_2  
0     0     0     1       0       1       0       0        1        0        0  
1     0     1     0       0       0       1       0        1        0        0  
2     1     0     0       0       0       1       0        0        0        1  
3     1     0     0       0       0       1       0        0        0        1  
4     0     0     0       0       0       1       0        0        0        1  

[5 rows x 22 columns]

In [6]:
# the 'target' row values will be stored in a seperate valiable to be used as labels for supervised learning approaches.
data_labels = df.target.values
X = df.drop(['target'], axis = 1)

In [7]:
# normalizing our data using min-max scaling

# create a scaler object
scaler = MinMaxScaler()

# fitting and transforming the data
normalized_data = pd.DataFrame(scaler.fit_transform(X), columns=X.columns)


In [8]:
# a sample of our normalized data
normalized_data.head()

age  sex  trestbps      chol  fbs  restecg   thalach  exang   oldpeak  \
0  0.708333  1.0  0.481132  0.244292  1.0      0.0  0.603053    0.0  0.370968   
1  0.166667  1.0  0.339623  0.283105  0.0      0.5  0.885496    0.0  0.564516   
2  0.250000  0.0  0.339623  0.178082  0.0      0.0  0.770992    0.0  0.225806   
3  0.562500  1.0  0.245283  0.251142  0.0      0.5  0.816794    0.0  0.129032   
4  0.583333  0.0  0.245283  0.520548  0.0      0.5  0.702290    1.0  0.096774   

    ca  ...  cp_1  cp_2  cp_3  thal_0  thal_1  thal_2  thal_3  slope_0  \
0  0.0  ...   0.0   0.0   1.0     0.0     1.0     0.0     0.0      1.0   
1  0.0  ...   0.0   1.0   0.0     0.0     0.0     1.0     0.0      1.0   
2  0.0  ...   1.0   0.0   0.0     0.0     0.0     1.0     0.0      0.0   
3  0.0  ...   1.0   0.0   0.0     0.0     0.0     1.0     0.0      0.0   
4  0.0  ...   0.0   0.0   0.0     0.0     0.0     1.0     0.0      0.0   

   slope_1  slope_2  
0      0.0      0.0  
1      0.0      0.0  
2      0.0      1.0  
3      0.0      1.0  
4      0.0      1.0  

[5 rows x 21 columns]

In [9]:
# now we need to split the data. 80% will be used for training and other 20% will be kept for testing.
x_train, x_test, y_train, y_test = train_test_split(normalized_data, data_labels, test_size = 0.2,random_state=0)

In [10]:
# We will use logistic regression method in this part to estimate discrete value based on a given set of categorical independent variables

# creating the model
logistic_regression_model = LogisticRegression()
# fitting the model
logistic_regression_model.fit(x_train, y_train)
# testing and calculating accuracy
accuracy = logistic_regression_model.score(x_test, y_test)*100

print("Logistic Regression Test Accuracy {}".format(accuracy))

Logistic Regression Test Accuracy 86.88524590163934
